pageserver/src/tenant/mgr.rs

```diff
@@ -1101,101 +1313,313 @@ pub(crate) async fn attach_tenant(
-    #[error("tenant {0} already exists, state: {1:?}")]
-    TenantAlreadyExists(TenantId, TenantState),
-    #[error("tenant {0} already exists in secondary state")]
-    TenantExistsSecondary(TenantId),
-    #[error(transparent)]
-    Other(#[from] anyhow::Error),
 }
 
-/// Give the given closure access to the tenants map entry for the given `tenant_id`, iff that
-/// entry is vacant. The closure is responsible for creating the tenant object and inserting
-/// it into the tenants map through the vacnt entry that it receives as argument.
-///
-/// NB: the closure should return quickly because the current implementation of tenants map
-/// serializes access through an `RwLock`.
-async fn tenant_map_insert<F, R>(
+/// Guards a particular tenant_id's content in the TenantsMap.  While this
+/// structure exists, the TenantsMap will contain a [`TenantSlot::InProgress`]
+/// for this tenant, which acts as a marker for any operations targeting
+/// this tenant to retry later, or wait for the InProgress state to end.
+pub struct SlotGuard {
     tenant_id: TenantId,
-    insert_fn: F,
-) -> Result<Arc<Tenant>, TenantMapInsertError>
-where
-    F: FnOnce() -> R,
-    R: std::future::Future<Output = anyhow::Result<Arc<Tenant>>>,
-{
-    let mut guard = TENANTS.write().await;
-    let m = match &mut *guard {
-        TenantsMap::Initializing => return Err(TenantMapInsertError::StillInitializing),
-        TenantsMap::ShuttingDown(_) => return Err(TenantMapInsertError::ShuttingDown),
-        TenantsMap::Open(m) => m,
-    };
-    match m.entry(tenant_id) {
-        hash_map::Entry::Occupied(e) => match e.get() {
-            TenantSlot::Attached(t) => Err(TenantMapInsertError::TenantAlreadyExists(
-                tenant_id,
-                t.current_state(),
-            )),
-            TenantSlot::Secondary => Err(TenantMapInsertError::TenantExistsSecondary(tenant_id)),
-        },
-        hash_map::Entry::Vacant(v) => match insert_fn().await {
-            Ok(tenant) => {
-                v.insert(TenantSlot::Attached(tenant.clone()));
-                Ok(tenant)
+    old_value: Option<TenantSlot>,
+    upserted: bool,
+
+    /// [`TenantSlot::InProgress`] carries the corresponding Barrier: it will
+    /// release any waiters as soon as this SlotGuard is dropped.
+    _completion: utils::completion::Completion,
+}
+
+unsafe impl Send for SlotGuard {}
+unsafe impl Sync for SlotGuard {}
+
+impl SlotGuard {
+    fn new(
+        tenant_id: TenantId,
+        old_value: Option<TenantSlot>,
+        completion: utils::completion::Completion,
+    ) -> Self {
+        Self {
+            tenant_id,
+            old_value,
+            upserted: false,
+            _completion: completion,
+        }
+    }
+
+    /// Take any value that was present in the slot before we acquired ownership
+    /// of it: in state transitions, this will be the old state.
+    fn take_value(&mut self) -> Option<TenantSlot> {
+        self.old_value.take()
+    }
```

Comment:
Would be nice to
1. panic in Drop if we didn't provide a replacement through `upsert`
2. ensure via type system that this can be called only once, i.e., statically ensure all calls return `Some()`

Reply:
I've played with making this one stricter by adding an explicit delete() and asserting in Drop that either delete or upsert was called.  The trouble is that it breaks the ergonomics of `?` in functions that hold a SlotGuard: they must avoid using `?` because if they hit an error then their slotguard gets illegally dropped without being written.

The kinds of errrors that we hit while holding a SlotGuard are generally very unexpected things, like failing to write a config file, where ultimately we'll probably make those infallible via fatal local io errors, and the current behavior in this PR of dropping the whole slot is a good one -- e.g. we were trying to attach a tenant and something weird went wrong, so we drop the slot.

Reply:
I'm worried about dropping a tenant's slot before we're sure that all activity for that tenant has stopped.

This is kind of related to the Gate PR.

If we had async Drop, we could make the SlotGuard close the gate and wait for existing activity to cede.

Without it, IMO the SlotGuard should assert that the gate is closed and activity has ceded.